### PR TITLE
Add first working implementation

### PR DESCRIPTION
**Pull request by katcipis**
Still incipient :-)

### CODE REVIEW

File: mdtoc.go [hunk 1]
```diff
@@ -1,7 +1,71 @@
...
+	var original bytes.Buffer
+	var wroteHeader bool
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		_, err := original.Write([]byte(line + "\n"))
+		if err != nil {
+			return err
+		}
+		parsed := strings.Split(line, "#")
```
Comment (i4ki) [inline mdtoc.go]: I found this very obscure way of doing.

Why not something like:

```go
func parseHeader(line string) (int, string) {
	var i int
	for i = 0; i < len(line); i++ {
            if line[i] != '#' {
                 break
            }
	}

	return i, line[i:]
}

level, title := parseHeader("## Introduction")
```

Maybe adding `TrimRight` to remove leading '#' in the end of title (by spec) and `TrimSpace`.

**Review by i4ki — COMMENTED** [review]
Looks good. Just add a possible alternative implementation for parseHeader. Feel free to ignore ;-)

File: mdtoc.go (hunk 1 above)
Comment (katcipis) [reply mdtoc.go]: I appreciate the feedback, but to be honest I'm failing to find using string.Split to be obscure and looping manually byte per byte to be better. Also, not sure if this works properly if there is special characters on the string (like a rune with 16 bits that the first byte matches exactly with the ASCII/utf-8 byte for #).

The trimming makes sense, I'm not covering a lot of corner cases yet :+1:

Comment (i4ki) [reply mdtoc.go]: It was a simple example. To get runes, just iterate over the string with `for Idx, r := range line {}` then golang handles that automagically for you. You can make it explicit importing `utf8` and using `DecodeRuneInString`.

It's obscure because you're spliting the entire line only for counting.. and then to count you use a `for`  loop anyway in a different function... This could be solved one way with a simple for in a proper 'parseHeader' function.

Think: you slice the line by adjacent runes. Why? You can just count...

Comment (katcipis) [reply mdtoc.go]: I changed some stuff to add some corner cases from markdown, now I use the splitted string and I split by spaces. It is a little odd but it seems that spaces are required on markdown headers, or they are not headers at all. This way I can easily separate the header marker from the rest of the header. Take a look, it looks like crap. To be honest I'm focusing more on handling a lot of corner cases on the tests and less in the algorithm...it is pretty crappy :-)

**Comment by katcipis** [comment]
Seems good enough to start working :-)

File: mdtoc.go [hunk 2]
```diff
@@ -1,7 +1,192 @@
...
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.TrimSpace(line) == headerStart {
+			err := skipUntil(scanner, func(l string) bool {
+				return strings.TrimSpace(l) == headerEnd
+			})
+			if err != nil {
+				return fmt.Errorf("error removing headers(corrupted headers?): %s", err)
+			}
+			skipUntil(scanner, func(l string) bool { return l != "" })
```
Comment (i4ki) [inline mdtoc.go]: you're not verifying the err in case of entire content was skipped. Is this intentional?

File: mdtoc.go [hunk 3]
```diff
@@ -1,7 +1,192 @@
...
+		writeOutput("\n\n")
+	}
+
+	writeOutput(original.String())
+	return writeErr
+}
+
+func GenerateFromFile(inputpath string, output io.Writer) error {
+	file, err := os.Open(inputpath)
+	defer file.Close()
```
Comment (i4ki) [inline mdtoc.go]: defer must be placed after error handling and you're not checking for close errors.

File: mdtoc.go [hunk 4]
```diff
@@ -1,7 +1,192 @@
...
+
+func GenerateInPlace(inputpath string) error {
+	var output bytes.Buffer
+	err := GenerateFromFile(inputpath, &output)
+	if err != nil {
+		return err
+	}
+
+	file, err := os.Create(inputpath)
+	defer file.Close()
```
Comment (i4ki) [inline mdtoc.go]: same here

File: mdtoc_test.go [hunk 5]
```diff
@@ -35,19 +55,184 @@ func TestTOC(t *testing.T) {
...
+					"\nexpected:\n%q\ngot:\n%q\n\n",
 					want,
 					got,
 				)
 			}
 		})
 	}
 }
+
+type fakeReadWriter struct {
```
Comment (i4ki) [inline mdtoc_test.go:152]: top!

File: mdtoc.go [hunk 6]
```diff
@@ -1,7 +1,192 @@
...
+	}
+
+	var original bytes.Buffer
+	var isCodeSection bool
+	var wroteHeader bool
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.TrimSpace(line) == headerStart {
+			err := skipUntil(scanner, func(l string) bool {
```
Comment (i4ki) [inline mdtoc.go:109]: this logic could be simplified (avoiding nested anonymous functions) if `skipUntil` receive the string to stop at. Every use of skipUntil just compare string, no special use case for a function.
But this is up to you, only a minor comment.

**Review by i4ki — CHANGES_REQUESTED** [review]
looks good. I made only minor comments.

**Review by vitorarins — APPROVED** [review]
LGTM

File: mdtoc.go (hunk 2 above)
Comment (katcipis) [reply mdtoc.go]: No, missed the case where there was only the headers and nothing else

File: mdtoc.go (hunk 3 above)
Comment (katcipis) [reply mdtoc.go]: indeed :-)

File: mdtoc.go (hunk 6 above)
Comment (katcipis) [reply mdtoc.go:109]: Yeah, I have the case where I trim the spaces:

```
			err := skipUntil(scanner, func(l string) bool {
				return strings.TrimSpace(l) == headerEnd
			})
```

I could do a skipper that accepts just a string and always trim spaces, but if a document has the header and some lines with spaces only I would destroy those lines. Right now I'm just removing all newlines until I get to the text, but considering leaving spaces.

**Comment by katcipis** [comment]
Fixed the stupid mistakes, thanks @tiago4orion
